listenbrainz_spark/stats/common/listening_activity.py

```diff
@@ -0,0 +1,139 @@
+    latest_listen_ts = get_latest_listen_ts()
+    if stats_range == "all_time":
+        # all_time stats range is easy, just return time from LASTFM founding
+        # to the latest listen we have in spark
+        from_date = datetime(LAST_FM_FOUNDING_YEAR, 1, 1)
+        to_date = latest_listen_ts
+        # compute listening activity on annual basis
+        step = relativedelta(years=+1)
+        date_format = "%Y"
+        return from_date, to_date, step, date_format
+
+    # following are "last" week/month/year stats, here we want the stats of the
+    # previous week/month/year and *not* from 7 days ago to today so on.
+
+    # If we had used datetime.now or date.today here and the data in spark
+    # became outdated due to some reason, empty stats would be sent to LB.
+    # Hence, we use get_latest_listen_ts to get the time of the latest listen
+    # in spark and so that instead of no stats, outdated stats are calculated.
+    latest_listen_date = latest_listen_ts.date()
+
+    # from_offset: this is applied to the latest_listen_date to get from_date
+    # to_offset: this is applied to from_date to get to_date
+    if stats_range == "week":
+        from_offset = relativedelta(weeks=-2, weekday=MO(-1))
+        to_offset = relativedelta(weeks=+2)
+        # compute listening activity for each day, include weekday in date format
+        step = relativedelta(days=+1)
+        date_format = "%A %d %B %Y"
+    elif stats_range == "month":
+        from_offset = relativedelta(months=-2, day=1)  # start of the previous to previous month
+        to_offset = relativedelta(months=+2)
+        # compute listening activity for each day but no weekday
+        step = relativedelta(days=+1)
+        date_format = "%d %B %Y"
+    elif stats_range == "quarter":
+        from_offset = get_two_quarters_ago_offset(latest_listen_date)
+        to_offset = relativedelta(months=+6)
+        # compute listening activity for each week with date format as day
+        step = relativedelta(weeks=+1)
+        date_format = "%d %B %Y"
+    elif stats_range == "half_yearly":
+        from_offset = _get_half_year_offset(latest_listen_date)
+        to_offset = relativedelta(months=+12)
+        step = relativedelta(months=+1)
+        date_format = "%B %Y"
+    else:  # year
+        from_offset = relativedelta(years=-2, month=1, day=1)  # start of the previous to previous year
+        to_offset = relativedelta(years=+2)
+        step = relativedelta(months=+1)
+        # compute listening activity for each month
+        date_format = "%B %Y"
+
+    from_date = latest_listen_date + from_offset
+    to_date = from_date + to_offset
+
+    # set time to 00:00
+    from_date = datetime.combine(from_date, time.min)
+    to_date = datetime.combine(to_date, time.min)
+
+    return from_date, to_date, step, date_format
+
+
+def setup_time_range(stats_range: str) -> Tuple[datetime, datetime]:
```

Comment:
docstring missing.

Reply:
This function is rather critical to how this PR works, so I think more docs are deffo in order.